xfer/sender.go

```diff
@@ -0,0 +1,248 @@
+package xfer
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"sync"
+
+	"github.com/weaveworks/scope/common/sanitize"
+)
+
+// Sender takes ownership of ReadClosers and tries to send them to a remote
+// receiver. Senders guarantee that the ReadCloser will be closed, even if it
+// isn't successfully sent.
+type Sender interface {
+	Send(io.ReadCloser) // must never block or fail
+	Stop()
+}
+
+// HTTPSender implements Sender by sending the most recent ReadCloser to a
+// remote target in the body of an HTTP POST. If the remote target is broken
+// or slow, HTTPSender will drop (and close) older ReadClosers in favor of
+// newer ones.
+type HTTPSender struct {
+	target    string
+	authToken string
+	probeID   string
+	recv      chan io.ReadCloser // from clients
+	send      chan io.ReadCloser // to remote
+	quit      chan struct{}
+}
+
+// NewHTTPSender returns a usable HTTP-based Sender.
+func NewHTTPSender(target, authToken, probeID string) Sender {
+	s := &HTTPSender{
+		target:    sanitize.URL("http://", AppPort, "/api/report")(target),
+		authToken: authToken,
+		probeID:   probeID,
+		recv:      make(chan io.ReadCloser),
+		send:      make(chan io.ReadCloser),
+		quit:      make(chan struct{}),
+	}
+	go s.recvLoop()
+	go s.sendLoop()
```

Comment:
Its not clear to me why we need these two go routines here; I think this can be achieved with a non blocking put to s.recv, which closes the rc if it fails.